### DISCUSSION TITLE
Transform component - option to disable inheriting parent's transform at runtime

### DISCUSSION THREAD

**Discussion by greerdv** [o3de/o3de#14416] [Ideas]
I'd like to add a setting to the transform component which allows turning off the inheritance of transform information from a parent (at runtime; you would probably still want the inheritance behaviour at edit time). The setting would be available via the transform interface, and would allow other components to switch off that behaviour if they will be responsible for controlling an entity's transform. 

Certain components are intended to control the transform of an entity, such as:
- Rigid body component
- Actor
- Attachment (a component which connects an entity to a joint on an actor e.g. so you can pick up an item and have it move with the character's hand)

If an entity is a child of another entity, then its transform will also update when the parent moves, and the two transform updates can end up fighting. This leads to a couple of problems:
- If components like the ones mentioned above are on a child entity, it will likely lead to unexpected behaviour, the cause of which is not obvious, hard to diagnose and generally not documented.
- The current easiest way to avoid problems is to only put entities with affected components at root level, but that itself has some problems:
    - There's no guidance at the moment to know that's what you need to do, either in documentation or UI.
    - It puts a somewhat arbitrary restriction on how you organize your entities. There are a lot of situations where it's natural from an organizational perspective to make entities children. Forcing those entities to live at root level could also lead to very flat hierarchies in the entity outliner which are hard to work with because they don't have a nice hierarchical structure.

Another use case would be making a camera attached to a character by a spring. In this case it would be natural to make the camera a child of the character, but currently that would mean it will be moved both by the movement of the parent and by the spring.

This would also support some potential new physics features such as articulations and vehicles, where it is desirable for the entity hierarchy to accurately represent the hierarchy of the structure being simulated. 

This change should be invisible to the majority of users - the components themselves will handle disabling the inheritance and no action will be required by users.

**Comment by tkothadev** [comment]
Would there be the ability to toggle this behavior at runtime?

**Reply by greerdv** [reply]
Yes. the entire point is to make it possible to toggle it at runtime.

**Comment by rainbj** [comment]
Hi @greerdv please keep me looped into this conversation. I think the idea of when we turn this on and off is important, and where that switch happens will be important. Thanks!

**Reply by greerdv** [reply]
Sure thing. You're right - some care will be required around when it is switched on and off. One interesting example would be something like a character holding a weapon (so it's an attachment, which turns off transform inheritance because the transform is linked to the animation). If the character drops the weapon, we might want to transition to it being a rigid body (which would also turn off transform inheritance, but you'd need to be careful about handling the transition).

**Reply by SergeyAMZN** [reply]
Something to keep in mind there's hierarchy system in Multiplayer gem for handling things like characters holding a weapon. 
https://www.o3de.org/docs/user-guide/networking/multiplayer/hierarchy/

**Comment by AMZN-daimini** [comment]
I'm not sure I agree with the addition of this setting.

To my understanding, the sole purpose of the Transform hierarchy is to make it so that the children entities' transforms are directly impacted by the transforms of the ancestors. That is why we have a local position/rotation/scale, which is the transform for the individual entity, and a world position/rotation/scale, which is generated by concatenating the local transform with all the local transforms of all the ancestors up to the root.
The setting you propose would break this rule... but that's literally the only rule defining the hierarchical bond between the entities. I don't think it's a "somewhat arbitrary restriction" when it's established that it's the only thing defining the hierarchical relationship.

As such, there is already a way to "turn off the inheritance of transform information from a parent at runtime", which is setting the transform to `EntityId()` to unparent from the Entity.
Also, as far as I can remember, this is how things behave in all other major engines as well. Which is not necessarily a reason to rule out anything that strays away from the norm, mind you, but still makes me think that it doesn't really prevent any workflow or behavior from being achievable.

---

If anything, reading your proposal makes me think whether we could have a clearer definition of which Components want to take control of the transform of the entity - it should not be possible for two components to do it at the same time, since it will result in undefined behavior, as you mentioned.

So a different approach in making things clearer for the user could be:

- Defining a service provided by these types of components that take control of the transform (for example, "TransformController")
- Make it so that all components you list provide this service and are incompatible with it, so that only one at a time can be active
- In the runtime implementation of these components, have them automatically unparent the entity so it's at the root.
- Extra - somehow signal that the component is going to break the parent-child bond at runtime, either with an icon or a warning in the component card.

---

TLDR: In general I don't really like the proposed solution but I'm open to discussing the problem :)

**Reply by greerdv** [reply]
@AMZN-daimini I don't really like the automatic unparenting idea - I think it's better for the runtime hierarchy to agree with the editor hierarchy. Changing the hierarchy at runtime seems to me like an even more dramatic intervention than changing the rules about inheriting.

I like the idea of making it so no two components can simultaneously control the transform, but I don't think the transform compatibility system is the best way of doing it since it would prevent those components from coexisting on an entity at all. The example above where you want to have something which can be held by a character (an attachment) or could be dropped and become a rigid body would not be possible without destroying one entity and spawning another because the rigid body component would be incompatible with the attachment component. Currently it would be possible to manage the attach/detach and enable/disable physics calls. Although having said that, it's also currently possible to get yourself into a bad situation by simultaneously enabling physics and attaching to an actor (and that would still be possible under my proposed change, so I guess that's something else for me to think about).

I guess my way of looking at things is that by e.g. putting a rigid body component on an entity and activating physics, you are already saying you want its transform to be controlled differently and to me it seems like ceasing to inherit from your parent is not a giant leap beyond that.

Another question is whether users consider the entity hierarchy to be a transform hierarchy (which is what is literally true) or if their mental model treats it as a conceptual/logical/organizational hierarchy. Or maybe it's all of those things and it's doing too many jobs.

**Reply by greerdv** [reply]
Another thing worth mentioning is that in a lot of the cases I'm thinking about, my change is actually maintaining the idea of the parent influencing the child, it's just not doing it via the transform bus. For example, if you have an attachment as a child of an actor, it's effectively getting moved by its parent, it's just not doing it in the somewhat unnuanced way that the transform hierarchy does it.

**Reply by AMZN-daimini** [reply]
> Another thing worth mentioning is that in a lot of the cases I'm thinking about, my change is actually maintaining the idea of the parent influencing the child, it's just not doing it via the transform bus. For example, if you have an attachment as a child of an actor, it's effectively getting moved by its parent, it's just not doing it in the somewhat unnuanced way that the transform hierarchy does it.

I thought the Attachment component had an EntityId field to determine what it was attached to. If it's using the Transform parent, then I think it's a flawed design cause it's giving the same property two different meanings.

> The example above where you want to have something which can be held by a character (an attachment) or could be dropped and become a rigid body would not be possible without destroying one entity and spawning another because the rigid body component would be incompatible with the attachment component.

Regarding this issue... I would argue that those two are indeed two different object. A weapon used by the player that can be attached to his hand is a different object from a physical rigidbody laying on the ground and reacting to being kicked around. At that point, whether they are replacing each other via destroy/spawn, or they are two siblings under a parent entity that get enabled/disabled alternatively, they should indeed be two separate objects in my opinion.

**Reply by greerdv** [reply]
You're correct about how the attachment component works, but putting it on a child will not work. I don't think that's an unnatural thing to do though and I have seen people do it and get confused about the weird behaviour.

**Reply by greerdv** [reply]
And destroying an entity and creating another one seems to me like an awful lot of work compared to calling Detach on one component and EnablePhysics on another component.

**Reply by AMZN-daimini** [reply]
The creating and destroying part was about incompatible components, just "Detach on one component and EnablePhysics on another component" is still equivalent to unparenting the entity and calling EnablePhysics.

I think specific use cases are actually pretty useful here. In the case you described in a comment above, for example...

> One interesting example would be something like a character holding a weapon (so it's an attachment, which turns off transform inheritance because the transform is linked to the animation). If the character drops the weapon, we might want to transition to it being a rigid body (which would also turn off transform inheritance, but you'd need to be careful about handling the transition).

Dropping the gun and leaving it parented to the player, while removing it from the constraints of the transform hierarchy via the proposed toggle, sounds like the biggest pitfall of this system. If the object is dropped by the player, it makes sense for it to be unparented. The proposed solutions, IMO, results in an even more convoluted system where parenting becomes irrelevant and if anything the organizational angle is even less clear.

**Reply by greerdv** [reply]
Yeah, it would probably make sense to unparent the gun in that case. But my point is that if the character starts with the gun, it might be quite natural to have the gun a child of the character in the editor.

There are other situations like vehicles, where it would be nice from an organizational perspective to have e.g. wheel entities which are children of the vehicle. But at runtime the wheels will be controlled by the vehicle simulation to handle suspension etc. Under the current system, all the parts of the vehicle would just have to be in a big flat list, and the hierarchical structure of the vehicle can only be deciphered by digging into the links between entities.

**Reply by AMZN-daimini** [reply]
Do the physics simulations affect the world transform of the object directly? Couldn't they calculate the world position, rotation etc but then set it to the entity as local values instead?
Cause the more we discuss the more it feels like a part of these issues may be due to misunderstandings or disagreements on the implementation side, and not really that the architecture makes any specific operation impossible.

**Reply by greerdv** [reply]
Even if you do the physics updates as local transform updates, you still might end up doing two updates if you have a parented entity. For example, if you have two bodies connected by a joint and one body is the child of the other (which again I think is quite a natural way to represent things in the editor), then the physics system would move both bodies, then you'd update the parent, update the child because the parent has moved, then update the local transform of the child. If it's a big tree of entities you will end up cascading a tonne of extra changes.

**Comment by moraaar** [comment]
@greerdv and I talked more about this yesterday, this is a summary of what was discussed.

We contemplated 3 approaches for achieving the expected behaviour for a hierarchy like "vehicle and wheels" that are controlled by physics and therefore the transform component "gets in the way":

**Method 1: use empty entities to group entities**
This is probably the method that game developers are more used to to avoid the transform parenting issue. Basically empty entities (which are never moved) are used as parents but the relevant entities are not directly related.

Entity Outliner
+ Vehicle (empty entity)
    + Vehicle Body
    + Wheels (empty entity)
        + Wheel 1
        + Wheel 2
        + Wheel 3
        + Wheel 4

Notice how wheel 1, 2, 3 and 4 are not children of vehicle body.

Pros:
- No alterations to transform component in entities, so the rules are kept simple.

Cons:
- It's not a natural way to construct hierarchies. Game developers might be familiarized with this method but not other industries.
- Users will have to be educated into generating this hierarchy when certain components are used (joints, simulated rigid bodies, articulations, etc.)
- It's still not good for cases like a robotic arm that has one long line of connections, it will end up with many siblings, losing the sense of hierarchy in entity outliner.
- Still this doesn't protect the root parent from being moved and there is nothing suggesting you shouldn't do that. 

**Method 2: components disable inheriting parent's transform**
**This is the method proposed in this discussion.** This is transparent to the user, components internally will be able to tell the transform to disable inheritance. 

Entity Outliner
+ Vehicle Body
    + Wheel 1
    + Wheel 2
    + Wheel 3
    + Wheel 4

Notice how wheel 1, 2, 3 and 4 are children of vehicle body. Since wheel entities will have a component that makes the entity be controlled by simulation (for example a joint component or a simulated rigid body), that component will internally disable parent's transform so there is no issues with the parent entity "Vehicle Body".

Pros:
- More natural way to construct hierarchies.
- Works automatically with the components that are not compatible with having a parent changing the transforms, like physics (joints, simulated rigid bodies, etc.)
- Users doesn't have to worry about where they put the components inside a hierarchy.
- Game developers that still prefer to construct the hierarchies like in method 1 they still can.

Cons:
- Components are changing how entity hierarchy works under the hood without the users knowing. This is not precisely a bad thing, specially if the entity outliner would be very clear that this entity is different than the others (for example by looking different).
- Developers creating new components have one more thing to consider. Again, not necessarily a bad thing, but it makes the process a bit more complex and it's hard to predict how this fits into future components.

**Method 3: expose to users the option to disable inheriting parent's transform in the transform component**
In this method the transform component have the option to turn on/off (by default on) inheriting the parent's transform.

Entity Outliner
+ Vehicle Body
    + Wheel 1
    + Wheel 2
    + Wheel 3
    + Wheel 4

Same hierarchy as method 2, but the users have to manually disable the "inherit transform" option in all the entities.

Pros:
- Same natural way to construct hierarchies as in method 2.
- This method simply adds a new feature to the transform component, with clear rules for the users to use, it's not hidden behind components controlling it.
- Users have the control of the behaviour of the entity.
- Still allows the users to create the hierarchy of method 1 if they so desired. 

Cons:
- Since the user controls the option, they can potentially make mistakes using it in combination with certain components. This disadvantage might be mitigated with a clear UI making components incompatible the "inherit transform" option.
- Exposing the feature to users might be an overcomplicated solution if it's clear that components could handle that themselves.

**My feedback**
I incline towards Method 2 (what this discussion is proposing) because it improves the experience using any component that makes the entity to be controlled by physics (or any other component that needs it). It also allows to create more intuitive hierarchies in the entity outliner. But I think it's crucial the UI shows in the entity outliner that those entities are different, to make it clear they won't be affected by the parent's transform.

Method 2 is also a good approach to start with, adding the extra functionality to transform component without making it more complicated for users, it's all done under the hood and the components controls it. But if afterwards the feedback from users is that it's not clear enough or that they will want the control, then the option can easily be exposed to become Method 3.

Cheers

**Comment by greerdv** [comment]
This has created a fair amount of discussion, so I have created an RFC to make sure everyone has an opportunity to comment https://github.com/o3de/sig-simulation/issues/57
